File: src/chapters/interpolation/content/code/newton-eval.go
package main

import "fmt"

// Evaluate the Newton form by nested (Horner-like) multiplication.
func newtonEval(x, a []float64, t float64) float64 {
	p := a[len(a)-1]
	for k := len(a) - 2; k >= 0; k-- {
		p = p*(t-x[k]) + a[k]
	}
	return p
}

func main() {
	fmt.Println(newtonEval([]float64{-1, 1, 2, 3}, []float64{-3, 2, 0, 3}, 0)) // 5
}

File: src/pages/practice/guides/matlab/newton-eval.m
function p = newton_eval(x, a, t)
% NEWTON_EVAL  Evaluate the Newton form by nested multiplication.
    p = a(end);
    for k = numel(a)-1:-1:1
        p = p*(t - x(k)) + a(k);
    end
end

% --- Demo ---
disp(newton_eval([-1 1 2 3], [-3 2 0 3], 0));   % 5
